kivy/input/providers/xinput2.pyx

```diff
@@ -0,0 +1,124 @@
+﻿"""
+Xinput2 input provider
+=========================
+"""
+
+cdef extern from "xinput2_core.c":
+    ctypedef struct TouchStruct:
+        int t_id, state
+        float x, y
+
+cdef extern int init(int window_id)
+cdef extern int idle()
+
+# Dummy class to bridge between the MotionEventProvider class and C
+class InputX11():
+
+    def __init__(self):
+        self.on_touch = None
+
+    def start(self, window_id, on_touch):
+        self.on_touch = on_touch
+        init(window_id)
+
+    def x11_idle(self):
+        idle()
+
+xinput = InputX11()
+
+# Create callback for touch events from C->Cython
+ctypedef int (*touch_cb_type)(TouchStruct *touch)
+cdef extern void x11_set_touch_callback(touch_cb_type callback)
+cdef int event_callback(TouchStruct *touch):
+    py_touch = {"id": touch.t_id,
+                "state": touch.state,
+                "x": touch.x,
+                "y": touch.y}
+    xinput.on_touch(py_touch)
+x11_set_touch_callback(event_callback)
+
+
+#######
+# Input provider
+__all__ = ('Xinput2EventProvider', 'Xinput2Event')
+
+from kivy.logger import Logger
+from collections import deque
+from kivy.base import EventLoop
+from kivy.input.provider import MotionEventProvider
+from kivy.input.factory import MotionEventFactory
+from kivy.input.motionevent import MotionEvent
+
+
+class Xinput2Event(MotionEvent):
+
+    def depack(self, args):
+        if args[0] is None:
+            return
+        self.profile = ('pos',)
+        width, height = EventLoop.window.system_size
+        rx = args[0] / float(width)
```

Comment:
Probably should use `width - 1` and `height - 1` when converting to relative coords. Same as in https://github.com/kivy/kivy/blob/master/kivy/input/providers/mouse.py#L275 . See https://github.com/kivy/kivy/pull/7425.

Reply:
I wonder if we should add a `normalize_pos` method to window that does this that can just be re-used everywhere, because otherwise it'd inevitably be forgotten in new providers.

Reply:
I agree for `normalize_pos` method.